indexing-service/src/main/java/org/apache/druid/indexing/common/task/KillUnusedSegmentsTask.java

```diff
@@ -300,6 +356,81 @@ private NavigableMap<DateTime, List<TaskLock>> getNonRevokedTaskLockMap(TaskActi
     return taskLockMap;
   }
 
+  private List<DataSegment> findUnreferencedSegments(
+      List<DataSegment> unusedSegments,
+      Map<String, String> upgradedFromSegmentIds,
+      TaskActionClient taskActionClient
+  ) throws IOException
+  {
+    if (unusedSegments.isEmpty() || upgradedFromSegmentIds.isEmpty()) {
+      return unusedSegments;
+    }
+
+    final Set<String> upgradedSegmentIds = new HashSet<>();
+    for (DataSegment segment : unusedSegments) {
+      final String id = segment.getId().toString();
+      upgradedSegmentIds.add(id);
+      if (upgradedFromSegmentIds.containsKey(id)) {
+        upgradedSegmentIds.add(upgradedFromSegmentIds.get(id));
+      }
+    }
+
+    final Map<String, Set<String>> upgradedToSegmentIds;
+    try {
+      upgradedToSegmentIds = taskActionClient.submit(
+          new RetrieveUpgradedToSegmentIdsAction(getDataSource(), upgradedSegmentIds)
+      ).getUpgradedToSegmentIds();
+    }
+    catch (Exception e) {
+      LOG.warn(
+          e,
+          "Could not retrieve UpgradedToSegmentsResponse using task action[retrieveUpgradedToSegmentIds]."
+          + " Overlord may be on an older version."
+      );
+      return unusedSegments;
+    }
+
+    final Set<String> existingSegmentIds
+        = taskActionClient.submit(new RetrieveSegmentsByIdAction(getDataSource(), upgradedSegmentIds))
```

Comment:
Why is this action needed?
Doesn't the new `UpgradedToSegmentIdsResponse` already cover this case?

See the comment here:
https://github.com/apache/druid/pull/16667#issuecomment-2205060939

Reply:
I have pushed changes with the new action specifications. However, the semantics for it seem wrong to me.
I think we shouldn't include a mapping from a segment to itself in this action for homogeneity.

Reply:
Sure, that works for me. Make sure the javadoc of the action captures exactly what is going to be included in the map.

Reply:
Done